new security guidelines & info (#54)
swift-server/sswg: security/package-maintainer-received-vulnerability-report.md

## Dealing with security vulnerability reports

> This information is intended for package authors who have packages listed on the [SSWG's package index](https://swift.org/server/#projects). If instead you found (or have heard of) a security vulnerability you'd like to report, please have a look over [here](contributor-found-vulnerability.md).

Reviewer: ```suggestion
This information is intended for package authors who have packages incubated by the SSWG and listed on the [SSWG's package index](https://swift.org/server/#projects). If instead you found (or have heard of) a security vulnerability you'd like to report, please have a look over [here](contributor-found-vulnerability.md).
```

---

The discovery of a security vulnerability in your code can be daunting, but it is part of the life of any software developer. So first of all, don't stress. If you need any help at any step of the way, please feel free to contact any of the SSWG members or the group as a whole at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org).

Here is a step-by-step guide on what to do with regards to the SSWG:

> 1. The SSWG is only a secondary point of contact which can broadcast information about vulnerabilities. The best way to start is to start addressing the vulnerability according the the security process of your own package.

Reviewer 2: ```suggestion
1. The SSWG is only a secondary point of contact which can broadcast information about vulnerabilities. The best way to start is to start addressing the vulnerability according to the security process of your own package.
```

2. As soon as you can -- but within 10 calendar days of discovering / receiving notification about the vulnerability -- please notify the SSWG about the vulnerability at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org). The SSWG will not disclose any information about this vulnerability and the emails can only be seen by the SSWG members listed on [swift.org/server](https://swift.org/server/) as well as the [Swift Core team](https://swift.org/community/#community-structure). If you prefer to share the vulnerability only with a smaller group, please feel free to reach out to any of the SSWG members individually.
3. After fixing the vulnerability, please promptly (within three days of releasing the fixed version) create a new Swift Forums post in the [Server > Security Updates](https://forums.swift.org/c/server/security-updates/50) category, linking to your own security advisory. The security advisory should contain at least which versions of what software are affected and how to update to an unaffected version.

> From [_graduated_](https://github.com/swift-server/sswg/blob/master/process/incubation.md#process-diagram) projects the SSWG typically expects that the whole process -- from the initial report/discovery to fixing and publishing the vulnerability -- is completed within 30 days. We do however acknowledge that certain types of vulnerabilities are either very complicated to address or a part of a "coordinated disclosure" which means that 30 days may not be enough. That is absolutely understandable, please however make sure to inform the SSWG (at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org)) about your anticipated timelines and any significant divergence from the plan.

Reviewer: ```suggestion 
[_graduated_](https://github.com/swift-server/sswg/blob/master/process/incubation.md#process-diagram) projects are expected to complete the whole process -- from the initial report/discovery to fixing and publishing the vulnerability -- within 30 days. We do however acknowledge that certain types of vulnerabilities are either very complicated to address or a part of a "coordinated disclosure" which means that 30 days may not be enough. That is absolutely understandable, please however make sure to inform the SSWG (at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org)) about your anticipated timelines and any significant divergence from the plan.
```

> A package maintainer's failure to report or address vulnerabilities may result in the SSWG publishing a security advisory, and could lead to retracting the project's status and listing it under the non-recommended projects list. In some cases, the SSWG may choose to find a technical contributor that can help resolve the security issues to minimize the impact. SSWG actions will be decided on a case by case basis and require a supermajority vote.

Reviewer: ```suggestion
A package maintainer's failure to report or address vulnerabilities may result in the SSWG publishing a security advisory, and could lead to retracting the project's status and listing it under the non-recommended projects list. In some cases, the SSWG may choose to find a technical contributor that can help resolve the security issues to minimize the impact on the ecosystem. SSWG actions will be decided on a case by case basis and require a supermajority vote.
```

> Project authors are also encouraged to make use of their source control system security features (for example: Github's "Security Advisories" and Gitlab's "Confidential Issues") to manage the vulnerabilities and inform their users.

Reviewer 3: ```suggestion
Project authors are also encouraged to make use of their source control system security features (for example: GitHub's "Security Advisories" and GitLab's "Confidential Issues") to manage the vulnerabilities and inform their users.
```

> Project authors are also encouraged to make use of their source control system security features (for example: GitHub's "Security Advisories" and GitLab's "Confidential Issues") to manage the vulnerabilities and inform their users.

Reviewer 2: ```suggestion
Project authors are also encouraged to make use of their source control system security features (for example: [GitHub's "Security Advisories"](https://docs.github.com/en/github/managing-security-vulnerabilities/about-github-security-advisories) and [GitLab's "Confidential Issues"](https://docs.gitlab.com/ee/user/project/issues/confidential_issues.html)) to manage the vulnerabilities and inform their users.
```

> Project authors are also encouraged to make use of their source control system security features (for example: [GitHub's "Security Advisories"](https://docs.github.com/en/github/managing-security-vulnerabilities/about-github-security-advisories) and [GitLab's "Confidential Issues"](https://docs.gitlab.com/ee/user/project/issues/confidential_issues.html)) to manage the vulnerabilities and inform their users.

Reviewer 4: I wonder if this shouldn't be a bit stronger than "encouraged" - to borrow RFC 2119-speak, not a `MUST` ("required"), but at least a `SHOULD` ("recommended"). "Encouraged" feels more like a `MAY` ("optional").

Author: @gwynne My thought here was that some projects may prefer to use other vulnerability management & notification systems. So I thought it's enough to "encourage" them to use the built-in ones in Github/Gitlab but if they have something else that works, that's cool too. To me the important part was that each project does have a process and defines it. WDYT?

Reviewer 5: I agree with this. There are multiple reasons why a repository/company may want to see alternative mechanisms; we shouldn't be as strong as `SHOULD` unless we can clearly articulate why alternative mechanisms would not be sufficient.

That said, we could add language around here to require that it must be easy for someone looking at a package to see its security advisories. This is obviously easier when using the source control system security features but doesn't rule out alternative mechanisms.
